github/messages.go

```diff
@@ -132,17 +137,37 @@ func messageMAC(signature string) ([]byte, func() hash.Hash, error) {
 //       // Process payload...
 //     }
 //
-func ValidatePayload(r *http.Request, secretKey []byte) (payload []byte, err error) {
-	payload, err = ioutil.ReadAll(r.Body)
-	if err != nil {
-		return nil, err
+func ValidatePayload(r *http.Request, secretKey []byte) ([]byte, error) {
+	// jsonPayload is the json webhook data
+	var jsonPayload []byte
+	// rawPayload is the raw body that GitHub uses to calculate the signature
+	var rawPayload []byte
+
+	if r.Header.Get("Content-Type") == "application/x-www-form-urlencoded" {
+		// If the content type is application/x-www-form-urlencoded
+		// the json payload will be under the "payload" form param
+		jsonPayload = []byte(r.FormValue(payloadFormParam))
+
+		// GitHub calculates the signature based on the query-escaped
+		// post body. In order to validate, we need to reconstruct the raw body.
+		rawPayload = []byte(fmt.Sprintf("%s=%s", payloadFormParam, url.QueryEscape(string(jsonPayload))))
+	} else {
+		defer r.Body.Close()
```

Comment:
I see.

However, this function is meant to be used on in HTTP handlers working on server HTTP requests, not client ones.

Specifically:

```Go
func (s *GitHubEventMonitor) ServeHTTP(w http.ResponseWriter, r *http.Request) {
    payload, err := github.ValidatePayload(r, s.webhookSecretKey)
    if err != nil { ... }
    // Process payload...
}
```

HTTP handlers do not need to close the response body.

See documentation here:

https://godoc.org/net/http#Request.Body

> For server requests the Request Body is always non-nil
> but will return EOF immediately when no body is present.
> **The Server will close the request body. The ServeHTTP
> Handler does not need to.**

So, the original behavior was correct, and you should revert that change.